frame/staking/src/offchain_election.rs

```diff
@@ -73,42 +88,108 @@ pub(crate) const DEFAULT_LONGEVITY: u64 = 25;
 /// Returns `Ok(())` if offchain worker should happen, `Err(reason)` otherwise.
 pub(crate) fn set_check_offchain_execution_status<T: Config>(
 	now: T::BlockNumber,
-) -> Result<(), &'static str> {
+	threshold: T::BlockNumber,
+) -> Result<(), OffchainElectionError> {
 	let storage = StorageValueRef::persistent(&OFFCHAIN_HEAD_DB);
-	let threshold = T::BlockNumber::from(OFFCHAIN_REPEAT);
 
 	let mutate_stat =
-		storage.mutate::<_, &'static str, _>(|maybe_head: Option<Option<T::BlockNumber>>| {
+		storage.mutate(|maybe_head: Option<Option<T::BlockNumber>>| {
 			match maybe_head {
-				Some(Some(head)) if now < head => Err("fork."),
+				Some(Some(head)) if now < head => Err(OffchainElectionError::Fork),
 				Some(Some(head)) if now >= head && now <= head + threshold => {
-					Err("recently executed.")
+					Err(OffchainElectionError::TooRecent)
 				}
 				Some(Some(head)) if now > head + threshold => {
-					// we can run again now. Write the new head.
+					// we can allow again now. Write the new head.
 					Ok(now)
 				}
 				_ => {
-					// value doesn't exists. Probably this node just booted up. Write, and run
+					// value doesn't exists. Probably this node just booted up. Write, and allow.
 					Ok(now)
 				}
 			}
-		});
+	});
 
+	crate::log!(trace, "attempting to acquire the OCW lock at {:?} = {:?}", now, mutate_stat);
 	match mutate_stat {
 		// all good
 		Ok(Ok(_)) => Ok(()),
 		// failed to write.
-		Ok(Err(_)) => Err("failed to write to offchain db."),
+		Ok(Err(_)) => Err(OffchainElectionError::DBWriteFailed),
 		// fork etc.
 		Err(why) => Err(why),
 	}
 }
 
+/// Storage path for the solution `call`.
+pub(crate) const OFFCHAIN_QUEUED_CALL: &[u8] = b"parity/staking-election/call";
+
+/// Save a given call OCW storage.
+pub(crate) fn save_solution<T: Config>(call: Call<T>) -> Result<(), OffchainElectionError> {
+	let storage = StorageValueRef::persistent(&OFFCHAIN_QUEUED_CALL);
+	// in all cases, just write the new value regardless of the the old one, if any.
+	let set_outcome = storage.mutate::<_, OffchainElectionError, _>(|_| Ok(call));
```

Comment:
It is the only way to set a value atomically, and we'd need that property in case multiple OCW threads are overlapping.

Reply:
Yeah I know, it is still a shitty interface. This could be a simple `set` function that does this automatically :D But, this is clearly nothing that is related to this pr ;)